Major wiki restructure, minor fixes, rewording (#1616)
hyprwm/hyprland-wiki: content/configuring/core/binds/_index.md

```lua
hl.bind("SUPER + TAB", hl.focus.workspace("e+1"))
hl.unbind("SUPER + Tab") -- this will NOT unbind
hl.unbind("SUPER + TAB") -- this will unbind
```

## XKB Remapping mod keys

You can customize the behavior of some modifier keys using `kb_options`.

To view all available options, run:

```sh
cat /usr/share/X11/xkb/rules/base.lst
```

{{% details title="Examples" closed="true" %}}

To remap Caps lock to Ctrl:

```lua
hl.config({
    input = {
        kb_options = "ctrl:nocaps"
    }
})
```

To swap Caps Lock and Escape:

```lua
hl.config({
    input =  {
        kb_options = "caps:swapescape"
    }
})
```

{{% /details %}}

## Binding quirks

<!-- TODO maybe there is a better place for this thing, but i couldnt see it -->

### Conditional binds resolution at "bind" time

When trying to create a bind with condition inside you almost always want to warp it in a function.

For example, this bind:

```lua
hl.bind("SUPER + L", hl.dsp.exec_cmd("foot", {float = not (hl.get_active_window().title == "foot") }) )
```

will be resolved to:

```lua
hl.bind("SUPER + L", hl.dsp.exec_cmd("foot", {float = false }) )

> -- or, depending on your focus when config was realoaded

Reviewer: ```suggestion
-- or, depending on your focus when config was reloaded
```

hl.bind("SUPER + L", hl.dsp.exec_cmd("foot", {float = false }) )
```
And it will stay like that until next config reload. 

> To change this behecaiour, dispatcher can be wrapped in a function:

Reviewer: ```suggestion
To change this behaviour, dispatcher can be wrapped in a function:
```

```lua
hl.bind("SUPER + L", function()
    local is_foot = hl.get_active_window().title == "foot"
    hl.dispatch( hl.dsp.exec_cmd("foot", {float = not is_foot }) )
end)
```

Now the condition will be evaluated on each call of the bind.

### Autoconsuming bind and return { ok = false }

> Hyprland exposes `ok` value to check result of a dispatcher. When combined with `auto_consumint` flag

Reviewer: ```suggestion
Hyprland exposes `ok` value to check result of a dispatcher. When combined with `auto_consuming` flag
```

it can be used to pass keys that are otherwise consumed by the bind.

For example:

```lua
hl.bind("p", function()
    local window = hl.get_active_window()
    if window and window.title == "some cool app" then
        hl.dispatch(hl.dsp.exec_cmd("another cool app"))
    else
        return { ok = false }
    end

> end, {auto_consumint = true})

Reviewer: ```suggestion
end, {auto_consuming = true})
```

Author: kill me

```

This bind will spawn `"another cool app"` if active window is `"some cool app"` and if

> active window is not `"some cool app"` it will pas `p` to that window.

Reviewer: ```suggestion
active window is not `"some cool app"` it will pass `p` to that window.
```

```lua
hl.bind("SUPER + TAB", hl.focus.workspace("e+1"))
hl.unbind("SUPER + Tab") -- this will NOT unbind
hl.unbind("SUPER + TAB") -- this will unbind
```

## XKB Remapping mod keys

You can customize the behavior of some modifier keys using `kb_options`.

To view all available options, run:

```sh
cat /usr/share/X11/xkb/rules/base.lst
```

{{% details title="Examples" closed="true" %}}

To remap Caps lock to Ctrl:

```lua
hl.config({
    input = {
        kb_options = "ctrl:nocaps"
    }
})
```

To swap Caps Lock and Escape:

```lua
hl.config({
    input =  {
        kb_options = "caps:swapescape"
    }
})
```

{{% /details %}}

## Binding quirks

<!-- TODO maybe there is a better place for this thing, but i couldnt see it -->

### Conditional binds resolution at "bind" time

When trying to create a bind with condition inside you almost always want to warp it in a function.

For example, this bind:

```lua
hl.bind("SUPER + L", hl.dsp.exec_cmd("foot", {float = not (hl.get_active_window().title == "foot") }) )
```

will be resolved to:

```lua
hl.bind("SUPER + L", hl.dsp.exec_cmd("foot", {float = false }) )
-- or, depending on your focus when config was reloaded

> hl.bind("SUPER + L", hl.dsp.exec_cmd("foot", {float = false }) )

Reviewer: Both lines are exactly the same. I guess you wanted to have it be false and true.
```suggestion
hl.bind("SUPER + L", hl.dsp.exec_cmd("foot", {float = true }) )
```

Author: damn it, cope-paste

> [!TIP]
> If you are unsure of what your key's name/keysym/keycode is, you can use [`wev`](https://github.com/jwrdegoede/wev) to find out.

## Basics

Basic syntax is:

```lua
hl.bind("keys", dispatcher or function(), { bind_flags })
```

The dispatcher list can be found in [Dispatchers](../dispatchers).
You can also supply a Lua function to your bind, if you prefer, and the bind will call it for you.

{{% details title="Examples" closed="true" %}}

```lua
-- bind SUPER + SHIFT + Q to open Firefox
hl.bind("SUPER + SHIFT + Q", hl.dsp.exec_cmd("firefox"))

-- bind to an anonymous function with logic inside
hl.bind("SUPER + SHIFT + X", function()
    -- more logic...
    hl.dispatch(hl.dsp.window.float())
end)
```

{{% /details %}}

## Uncommon syms / binding with a keycode

See the [xkbcommon-keysyms.h header](https://github.com/xkbcommon/libxkbcommon/blob/master/include/xkbcommon/xkbcommon-keysyms.h) for all the keysyms.
The name you should use is the segment after `XKB_KEY_`.

To bind a keycode, use the `code:` prefix before the key.
For example:

```lua
hl.bind("SUPER + code:28", hl.dsp.exec_cmd("amongus"))
```

> will bind SUPER + T, since T is keycode 28.

Reviewer 2: ```suggestion
will bind `SUPER + T`, since `T` is keycode 28.
```

## Binding modkeys only

<!-- NOTE: https://github.com/hyprwm/Hyprland/pull/15568 -->

To bind only a modkey, use the appropriate keysym name.
Usually it is a `MOD` key, suffixed with `_L` or `_R`.
When a keysym is used, the order in which keys are pressed matters, because they are not treated as modifiers anymore.

{{% details title="Examples" closed="true" %}}

```lua
-- bind `amongus` to left alt
hl.bind("Alt_L", hl.dsp.exec_cmd("amongus"))

-- bind left super with left ctrl to open kitty
hl.bind("Super_L + Alt_L", hl.dsp.exec_cmd("kitty"))

-- bind left ctrl and right ctrl
hl.bind("Ctrl_L + Ctrl_R", hl.dsp.exec_cmd("kitty"))
```

{{% /details %}}

## Multiple binds to one key

> [!WARNING]

> > The keybinds will be executed top to bottom, in the order they were written in.

Reviewer 2: ```suggestion
> The keybinds will be executed top to bottom, in the order they were written.
```

> You can trigger multiple actions with the same keybind by using a Lua lambda function, which can then execute multiple dispatcher:

Reviewer 2: ```suggestion
You can trigger multiple actions with the same keybind by using a Lua lambda function, which can then execute multiple dispatch calls:
```

Author: thats not right. we are using multiple dispatches in a function, not multiple functions in a dispatch

Reviewer 2: This is a rewording, not a rewrite.

Each hl.dispatch() is a dispatch call, and it can be used in several.

Nothing about multiple functions in a dispatch.

LAMBDA can be used in MULTIPLE dispatch calls, NOT lambda can be used multiple times in a dispatch call

Reviewer 2: ok misunderstood the original sorry. fixed wording.

```lua
-- To switch between windows in a floating workspace:
hl.bind("SUPER + Tab", function()
    -- Change focus to another window
    hl.dispatch(hl.dsp.window.cycle_next())
    -- Bring it to the top
    hl.dispatch(hl.dsp.window.bring_to_top())
end)
```

## Unbind

To unbind a key use `hl.unbind("key")` or assign it to a variable and use `:unbind()` or `:remove()` methods.
The key in `hl.unbind` is case-sensitive and must exactly match the case of the `hl.bind` you are unbinding.
Unbind will remove all occurances of the key that were registered before it.

{{% details title="Examples" closed="true" %}}

```lua
hl.bind("SUPER + TAB", hl.focus.workspace("e+1"))
hl.unbind("SUPER + Tab") -- this will NOT unbind
hl.unbind("SUPER + TAB") -- this will unbind

-- assign bind to "lft"
local lft = hl.bind("SUPER + L", hl.dsp.window.move({ direction = "left" }) )

-- unbind "SUPER + L"
lft:unbind()
-- or

> lft:remove()

Reviewer 2: ```suggestion
local moveLeft = hl.bind("SUPER + L", hl.dsp.window.move({ direction = "left" }) )

-- unbind "SUPER + L"
moveLeft:unbind()
-- or
moveLeft:remove()
```

It's jarring to read `lft`

```

{{% /details %}}

## XKB options

<!-- NOTE: maybe there is a better place for this thing, but i couldnt find it -->

To change the behavior of some keys, use `kb_options`.

To view all available options, run:

```sh
cat /usr/share/X11/xkb/rules/base.lst
```

{{% details title="Examples" closed="true" %}}

To remap Caps Lock to Ctrl:

```lua
hl.config({
    input = {
        kb_options = "ctrl:nocaps"
    }
})
```

To swap Caps Lock and Escape:

```lua
hl.config({
    input =  {
        kb_options = "caps:swapescape"
    }
})
```

{{% /details %}}

## Binding quirks

<!-- NOTE: maybe there is a better place for this thing, but i couldnt find it -->

### Conditional bind resolution at bind time

When trying to create a bind with a condition inside it, you almost always want to wrap it in a function.
For example, this bind:

```lua
hl.bind("SUPER + L", hl.dsp.exec_cmd("foot", { float = not (hl.get_active_window().title == "foot") }) )
```

will be resolved to:

```lua
hl.bind("SUPER + L", hl.dsp.exec_cmd("foot", { float = false }) )
-- or, depending on your focus when config was reloaded

> hl.bind("SUPER + L", hl.dsp.exec_cmd("foot", { float = false }) )

Reviewer 2: ```suggestion
hl.bind("SUPER + L", hl.dsp.exec_cmd("foot", { float = false }) )
-- or, depending on your focus when config was reloaded
hl.bind("SUPER + L", hl.dsp.exec_cmd("foot", { float = true }) )
```

identical lines

Author: i swear to Emperor, im somehow fucking up commit history
